cupyx/distributed/_store.py

```diff
@@ -0,0 +1,154 @@
+import multiprocessing
+import pickle
+import threading
+import socket
+import time
+
+
+class _BarrierImpl:
+    def __init__(self, world_size):
+        self._world_size = world_size
+        self._cvar = threading.Condition()
+
+    def __call__(self):
+        # Superlame implementation, should be improved
+        with self._cvar:
+            self._world_size -= 1
+            if self._world_size == 0:
+                self._cvar.notifyAll()
+            elif self._world_size > 0:
+                self._cvar.wait()
+
+
+class TCPStore:
+    # This is only used for initialization of nccl so we don't care
+    # too much about peformance
+    def __init__(self, world_size):
+        self.storage = {}
+        self._process = None
+        self._run = multiprocessing.Value('b', True)
+        self._world_size = world_size
+        # For implementing a barrier
+        self._lock = threading.Lock()
+        self._current_barrier = None
+
+    def __del__(self):
+        if self._process is not None:
+            with self._run.get_lock():
+                self._run.value = 0
+            self._process.join()
+
+    class Set:
+        def __init__(self, key, value):
+            self.key = key
+            self.value = value
+
+        def __call__(self, store):
+            store.storage[self.key] = self.value
+
+    class Get:
+        def __init__(self, key):
+            self.key = key
+
+        def __call__(self, store):
+            return store.storage[self.key]
+
+    class Exit:
+        def __call__(self, store):
+            store._run = False
+
+    class Barrier:
+        def __call__(self, store):
+            with store._lock:
+                if store._current_barrier is None:
+                    store._current_barrier = _BarrierImpl(store._world_size)
+            store._current_barrier()
+            # Once the barrier has been completed, just clean it
+            with store._lock:
+                store._current_barrier = None
+            return True
+
+    def _set_process(self, process):
+        self._process = process
+
+    def _process_request(self, c_socket):
+        with c_socket:
+            data = c_socket.recv(1024)
+            r = pickle.loads(data)(self)
```

Comment:
Security ...

Reply:
As discussed offline, if we implement a server socket in a product we need to take care of security things. I think the very basic KLV binary format is sufficient as we don't have to worry about protocol versioning.

Please also add a disclaimer in docstring that:
* use of this function launches the server
* the functionality is expected to be used in a trusted cluster environment and there is no authentication systems